build: config: image: add redesigned options for squashfs compression

Spudz76 opened the pull request:
This is an attempt at making SquashFS compression fully selectable, and enables all the initramfs compression options (previously selectable, but broken).  I wanted to try LZ4 since it might boot a few seconds faster and my device has no shortage of flash space, but nobody wired up the options, and the host-build of `mksquashfs4` was missing several compressors for no real reason (space not an issue on the compilation host).

* Add `tools/liblz4`, `tools/liblzo`, `tools/lzop` for the host toolchain, set `tools/squashfskit4` to use them all (and already existing zstd)
* Add config menus for overriding the default squashfs compression (unset follows board-default or otherwise behaves as it always did)
* Fix wiring for initramfs compression methods which were already in menuconfig but didn't really work

This is all untested past building and inspecting output on the compile-host ; awaiting serial cable before I can test it on real hardware for actually-boots status

Did not set myself as maintainer for the new tools packages since they are roughly based on the target packages of similar names.  Also did not copy maintainer from those sources, but IMHO it should just be the same maintainer.

mpratt14 commented:
every single tool you add should be a separate commit

also the tools should not be always enabled by default, but rather use the config you are adding to select them

Spudz76 commented:
I will split commits as suggested, and I should have done so in the first place.

However `zstd` is always included, therefore what I followed with the other compression types.  I will attempt to make them connect with the `TARGET_INITRAMFS_COMPRESSION_*` settings (but for consistency, will also map `zstd`...)

`zlib` and `xz` are also enabled always in `squashfskit4` and always built, even if it is unused by options... another precident I was generally following.  Those should also be set up to follow the selected options.

mpratt14 commented:
`zlib` and `xz` should for sure be always enabled, since other tools that are always needed are dependent on them somehow

I would just focus on the ones you're adding

dangowrt commented on tools/liblz4/Makefile:
```diff
@@ -0,0 +1,54 @@
+
+PKG_LICENSE:=BSD-2-Clause
+PKG_LICENSE_FILES:=LICENSE lib/LICENSE
+PKG_MAINTAINER:=
```
Strictly speaking this isn't a package, and most builds in `tools/` lack `PKG_MAINTAINER`. I suggest to drop `PKG_MAINTAINER` field instead of setting it to an empty string.

Spudz76 replied:
noted, thanks

I also tried to use `PKG_CONFIG_DEPENDS` which is apparently not wired for `tools`, figuring out a workaround for that.

Spudz76 commented:
> `zlib` and `xz` should for sure be always enabled, since other tools that are always needed are dependent on them somehow

If anything is truly dependent, then its deps should be wired up in the `tools/Makefile` like everything else.

I have already stripped them and things still generally work, but I have not tested with a fully clean/virgin tree (maybe the pre-existing tools are already previously installed and leftover in my `staging_dir`).

mpratt14 commented:
> I have already stripped them and things still generally work

careful with that concept

things can still "work" but that's not always the point, some tools are meant to be just "our copy" of whatever it is, and on many build systems you can install the same thing on the host machine, but that version of the tool differs between build systems. one of the purposes of the `xz` tool for example is version control to ensure every downloaded package that is an .xz image is unpacked the same way guaranteed and not subject to the differences between versions of `xz` provided by different OSes. Some host machines might have every tool installed already, but the build system is designed to be able to run on an almost completely "dry" system with only build-essential packages like git and make and autotools, etc. installed.

hauke commented on include/image-commands.mk:
```diff
@@ -277,6 +277,8 @@ define Build/initrd_compression
 	$(if $(CONFIG_TARGET_INITRAMFS_COMPRESSION_GZIP),.gzip) \
 	$(if $(CONFIG_TARGET_INITRAMFS_COMPRESSION_LZMA),.lzma) \
+	$(if $(CONFIG_TARGET_INITRAMFS_COMPRESSION_LZO),.lzo) \
+	$(if $(CONFIG_TARGET_INITRAMFS_COMPRESSION_LZ4),.lz4) \
```
Please add the initramfs fix into a separate commit too.

Spudz76 replied:
Way ahead of that, just haven't pushed in the last couple days.

hauke commented on config/Config-images.in:
```diff
@@ -153,6 +153,38 @@ menu "Target Images"
+				bool "xz"
+
+			config TARGET_SQUASHFS_COMPRESSION_ZSTD
+				depends on !LINUX_5_4 && !LINUX_4_19
```
OpenWrt master only supports kernel 5.10 and later. these dependencies are not needed.

Spudz76 replied:
Copied from the target package, which also has these, they should probably be nitpicked over there too?

hauke commented on config/Config-images.in:
```diff
@@ -153,6 +153,38 @@ menu "Target Images"
+				bool "lzo"
+
+			config TARGET_SQUASHFS_COMPRESSION_XZ
+				bool "xz"
```
You have to activate the compression standards also in the kernel configuration. Currently most of them are deactivated:
``
$ git grep SQUASHFS -- target/linux/generic/config-5.15
target/linux/generic/config-5.15:CONFIG_MTD_SPLIT_SQUASHFS_ROOT=y
target/linux/generic/config-5.15:CONFIG_SQUASHFS=y
target/linux/generic/config-5.15:# CONFIG_SQUASHFS_4K_DEVBLK_SIZE is not set
target/linux/generic/config-5.15:# CONFIG_SQUASHFS_DECOMP_MULTI is not set
target/linux/generic/config-5.15:CONFIG_SQUASHFS_DECOMP_MULTI_PERCPU=y
target/linux/generic/config-5.15:# CONFIG_SQUASHFS_DECOMP_SINGLE is not set
target/linux/generic/config-5.15:CONFIG_SQUASHFS_EMBEDDED=y
target/linux/generic/config-5.15:# CONFIG_SQUASHFS_FILE_CACHE is not set
target/linux/generic/config-5.15:CONFIG_SQUASHFS_FILE_DIRECT=y
target/linux/generic/config-5.15:CONFIG_SQUASHFS_FRAGMENT_CACHE_SIZE=3
target/linux/generic/config-5.15:# CONFIG_SQUASHFS_LZ4 is not set
target/linux/generic/config-5.15:# CONFIG_SQUASHFS_LZO is not set
target/linux/generic/config-5.15:# CONFIG_SQUASHFS_XATTR is not set
target/linux/generic/config-5.15:CONFIG_SQUASHFS_XZ=y
target/linux/generic/config-5.15:# CONFIG_SQUASHFS_ZLIB is not set
target/linux/generic/config-5.15:# CONFIG_SQUASHFS_ZSTD is not set
``

Spudz76 replied:
Again, all much cleaner in the code nobody can see yet

hauke commented:
Could you provide some sizes for the different compression standards? Maybe we should switch from XZ to zstd. The compression ratio in zstd can be configured

Spudz76 commented:
> Could you provide some sizes for the different compression standards please.
> 
> Maybe we should switch the default from XZ to zstd. The compression ratio in zstd can be configured.

Yes, I collected stats from various build (all the same rootfs content)
```
Filesystem size 7725990 bytes (7544.91 Kbytes / 7.37 Mbytes) Compression xz Block size 1048576
Filesystem size 7863310 bytes (7679.01 Kbytes / 7.50 Mbytes) Compression xz Block size 524288
Filesystem size 7996442 bytes (7809.03 Kbytes / 7.63 Mbytes) Compression xz Block size 262144
Filesystem size 8180238 bytes (7988.51 Kbytes / 7.80 Mbytes) Compression xz Block size 131072
Filesystem size 8637166 bytes (8434.73 Kbytes / 8.24 Mbytes) Compression zstd Block size 1048576
Filesystem size 8795350 bytes (8589.21 Kbytes / 8.39 Mbytes) Compression zstd Block size 524288
Filesystem size 8924443 bytes (8715.28 Kbytes / 8.51 Mbytes) Compression zstd Block size 262144
Filesystem size 9088083 bytes (8875.08 Kbytes / 8.67 Mbytes) Compression zstd Block size 131072
Filesystem size 9673987 bytes (9447.25 Kbytes / 9.23 Mbytes) Compression gzip Block size 1048576
Filesystem size 9679301 bytes (9452.44 Kbytes / 9.23 Mbytes) Compression gzip Block size 524288
Filesystem size 9696926 bytes (9469.65 Kbytes / 9.25 Mbytes) Compression gzip Block size 262144
Filesystem size 9735833 bytes (9507.65 Kbytes / 9.28 Mbytes) Compression gzip Block size 131072
Filesystem size 10532724 bytes (10285.86 Kbytes / 10.04 Mbytes) Compression lzo Block size 1048576
Filesystem size 10548119 bytes (10300.90 Kbytes / 10.06 Mbytes) Compression lzo Block size 524288
Filesystem size 10587339 bytes (10339.20 Kbytes / 10.10 Mbytes) Compression lzo Block size 262144
Filesystem size 10670590 bytes (10420.50 Kbytes / 10.18 Mbytes) Compression lzo Block size 131072
Filesystem size 11170138 bytes (10908.34 Kbytes / 10.65 Mbytes) Compression lz4 Block size 1048576
Filesystem size 11203138 bytes (10940.56 Kbytes / 10.68 Mbytes) Compression lz4 Block size 524288
Filesystem size 11276815 bytes (11012.51 Kbytes / 10.75 Mbytes) Compression lz4 Block size 262144
Filesystem size 11435621 bytes (11167.60 Kbytes / 10.91 Mbytes) Compression lz4 Block size 131072
Filesystem size 13879944 bytes (13554.63 Kbytes / 13.24 Mbytes) Compression lz4 Block size 4096
```

And, I collected live performance stats for some of the same builds (linksys EA7300v1 / ramips-mt7261):
```
target.lz4.0004:root@test:~# sync; echo 3 > /proc/sys/vm/drop_caches; time /bin/sh -c "find /rom -type f | sort | xargs -n1 cat > /dev/null"
target.lz4.0004:real    0m 12.55s
target.lz4.0004:user    0m 5.26s
target.lz4.0004:sys     0m 3.12s
target.lz4.0128:root@test:~# sync; echo 3 > /proc/sys/vm/drop_caches; time /bin/sh -c "find /rom -type f | sort | xargs -n1 cat > /dev/null"
target.lz4.0128:real    0m 14.02s
target.lz4.0128:user    0m 5.13s
target.lz4.0128:sys     0m 6.69s
target.lz4.0256:root@test:~# sync; echo 3 > /proc/sys/vm/drop_caches; time /bin/sh -c "find /rom -type f | sort | xargs -n1 cat > /dev/null"
target.lz4.0256:real    0m 11.50s
target.lz4.0256:user    0m 5.10s
target.lz4.0256:sys     0m 3.39s
target.lz4.0512:root@test:~# sync; echo 3 > /proc/sys/vm/drop_caches; time /bin/sh -c "find /rom -type f | sort | xargs -n1 cat > /dev/null"
target.lz4.0512:real    0m 11.61s
target.lz4.0512:user    0m 5.16s
target.lz4.0512:sys     0m 3.39s
target.lz4.1024:root@test:~# sync; echo 3 > /proc/sys/vm/drop_caches; time /bin/sh -c "find /rom -type f | sort | xargs -n1 cat > /dev/null"
target.lz4.1024:real    0m 11.09s
target.lz4.1024:user    0m 5.16s
target.lz4.1024:sys     0m 3.23s
target.xz.0128:root@test:~# sync; echo 3 > /proc/sys/vm/drop_caches; time /bin/sh -c "find /rom -type f | sort | xargs -n1 cat > /dev/null"
target.xz.0128:real    0m 13.58s
target.xz.0128:user    0m 4.95s
target.xz.0128:sys     0m 6.38s
target.xz.0256:root@test:~# sync; echo 3 > /proc/sys/vm/drop_caches; time /bin/sh -c "find /rom -type f | sort | xargs -n1 cat > /dev/null"
target.xz.0256:real    0m 13.85s
target.xz.0256:user    0m 4.98s
target.xz.0256:sys     0m 6.43s
target.xz.0512:root@test:~# sync; echo 3 > /proc/sys/vm/drop_caches; time /bin/sh -c "find /rom -type f | sort | xargs -n1 cat > /dev/null"
target.xz.0512:real    0m 13.93s
target.xz.0512:user    0m 5.09s
target.xz.0512:sys     0m 6.50s
target.xz.1024:root@test:~# sync; echo 3 > /proc/sys/vm/drop_caches; time /bin/sh -c "find /rom -type f | sort | xargs -n1 cat > /dev/null"
target.xz.1024:real    0m 13.04s
target.xz.1024:user    0m 5.02s
target.xz.1024:sys     0m 6.11s
target.zstd.0512:root@test:~# sync; echo 3 > /proc/sys/vm/drop_caches; time /bin/sh -c "find /rom -type f | sort | xargs -n1 cat > /dev/null"
target.zstd.0512:real    0m 11.30s
target.zstd.0512:user    0m 5.06s
target.zstd.0512:sys     0m 3.78s
target.zstd.1024:root@test:~# sync; echo 3 > /proc/sys/vm/drop_caches; time /bin/sh -c "find /rom -type f | sort | xargs -n1 cat > /dev/null"
target.zstd.1024:real    0m 10.86s
target.zstd.1024:user    0m 5.09s
target.zstd.1024:sys     0m 3.68s
```

ZSTD is indeed the winner overall (slightly larger than XZ but way faster and less CPU load), but LZ4 can be really fast and lowest CPU load, if you don't care as much about space.

LZMA1 support should also be fully dropped, and `squashfskit4` is abandoned while the real `squashfs-tools` is much newer (and has dropped LZMA1 and fixed a bunch of other things since `squashfskit4` forked from it...)

Spudz76 commented:
Also, force-pushed my current work, unclassified work in progress lumped in `WIP`

hauke commented:
There is work ongoing in #10967 to switch back to squashfs-tools.
I think you can configure the compression ratio in zstd. did you experiment with that?

Spudz76 commented:
Rebased to current `master` and fixed a build error and a missing autoload in the lz4 kmod changes.

Verified it now builds from a clean kernel tree, and the lz4hc compressor kmod exists and gets loaded.

Currently reorganizing the menuconfig stuff since I don't like how it works yet (and it needs a setting for which of the enabled compression methods is actually used in the image, currently it just uses whichever the last-enabled one is).

Also spent some time investigating the `squashfskit4` situation, and some work on making a compressor-agnostic `loader` so that we might choose `zstd-loader` or others rather than just good old `lzma-loader` and save like 200ms on booting.

Should I set this to Draft and does that block the constant re-testing cycles (that probably will crash anyway until this is polished)?

dangowrt commented on tools/Makefile:
```diff
@@ -56,10 +65,13 @@ $(curdir)/genext2fs/compile := $(curdir)/libtool/compile
 $(curdir)/gengetopt/compile := $(curdir)/libtool/compile
 $(curdir)/gmp/compile := $(curdir)/libtool/compile
 $(curdir)/isl/compile := $(curdir)/gmp/compile
+$(curdir)/liblzo/compile := $(curdir)/cmake/compile
```
This line representing a build-dependency of libbzo should go into the commit adding liblzo.

dangowrt commented on tools/Makefile:
```diff
@@ -56,10 +65,13 @@ $(curdir)/genext2fs/compile := $(curdir)/libtool/compile
 $(curdir)/libtool/compile := $(curdir)/automake/compile $(curdir)/missing-macros/compile
-$(curdir)/lzma-old/compile := $(curdir)/zlib/compile
 $(curdir)/llvm-bpf/compile := $(curdir)/cmake/compile
+$(curdir)/lz4/compile := $(curdir)/liblz4/compile
```
This line should go into the commit adding the lz4 tool

dangowrt commented on tools/Makefile:
```diff
@@ -56,10 +65,13 @@ $(curdir)/genext2fs/compile := $(curdir)/libtool/compile
 $(curdir)/llvm-bpf/compile := $(curdir)/cmake/compile
+$(curdir)/lz4/compile := $(curdir)/liblz4/compile
+$(curdir)/lzma-old/compile := $(curdir)/zlib/compile
+$(curdir)/lzop/compile := $(curdir)/cmake/compile $(curdir)/liblzo/compile
```
This line should go into the commit adding the lzop tool

dangowrt commented on tools/Makefile:
```diff
@@ -37,6 +44,8 @@ tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS)$(CONFIG_TARGET_tegra),y) += cbootimage
 tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS)$(CONFIG_USE_SPARSE),y) += sparse
 tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS)$(CONFIG_USE_LLVM_BUILD),y) += llvm-bpf
+tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS)$(BUILD_LZ4_TOOLS),y) += lz4
+tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS)$(BUILD_LZO_TOOLS),y) += lzop
```
This line should go into the commit adding the lzop tool

dangowrt commented on tools/Makefile:
```diff
@@ -37,6 +44,8 @@ tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS)$(CONFIG_TARGET_tegra),y) += cbootimage
 tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS)$(CONFIG_USES_MINOR),y) += kernel2minor
 tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS)$(CONFIG_USE_SPARSE),y) += sparse
 tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS)$(CONFIG_USE_LLVM_BUILD),y) += llvm-bpf
+tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS)$(BUILD_LZ4_TOOLS),y) += lz4
```
This line should go into the commit adding the lz4 tool

dangowrt reviewed (changes requested):
Please write commit descriptions for each commit, not just titlte and SoB (just in case you were not planning to still do that).
Apart from that and changes to [`tools/Makefile`](https://github.com/openwrt/openwrt/pull/11328/files#diff-e327741dae7d1efd73c66ab134d50ac2c56187d814c443bb55464272b8b44113) being in the wrong commit it already looks very nice up to this point.
Thank you for working on this and contributing!

Spudz76 commented:
improved commit descriptions, picked most of those oneliner cherries into the appropriate atomic commits; thank you for the tips/pointers/direction

the remaining lines have to be added to the matching commits without the `BUILD_LZ*_TOOLS` since those don't exist until a later commit, but agree they should get built when all-host-tools is active as of the package commit (even though literally nothing is wired up to depend on them until later)

Spudz76 reviewed (commented):
~~I believe [this single commit c9d2056](https://github.com/openwrt/openwrt/pull/11328/commits/c9d20563b7a171922015129499eb3b1e985863c8) is complete and could be merged, can that occur from here or should I make a totally separate branch and PR for it atomically?~~

Made a [separate PR#11358](https://github.com/openwrt/openwrt/pull/11358) for semi-unrelated, completely separable, and unused-here repair that seems to be unlikely to change in this PR.

tapper82 commented:
This looks interesting.

Spudz76 committed:
tools/squashfs4: adjust for squashfs option updates

enable all compressors, set `COMP_DEFAULT` to zstd, configuration can
easily override the default where zstd is not preferable

add PREFIX support to upstream Makefile via patch

define `Host/Uninstall` Makefile target, use default build args rather
than overriding `Host/Compile` and `Host/Install` targets

Signed-off-by: Tony Butler <spudz76@gmail.com>

Spudz76 committed:
build: config: image: add redesigned options for squashfs compression

add options within images for selecting squashfs compression type
using all possibilities provided in tools/squashfs4 and the kernel
(gzip, lz4, lzo, xz, zstd)

Also attempt to select the best default based on `SMALL_FLASH` and
`LOW_MEMORY_FOOTPRINT` leaning toward zstd for most platforms

If any compressors are disabled in the kernel config, those will not be
available and will not be selected.

Signed-off-by: Tony Butler <spudz76@gmail.com>

Spudz76 commented:
Did a whole bunch of cleanup and commit-splitting, now it actually looks almost good, and much less noise with unrelated things split out to:
* #11381 
* #11358 
* #11359 
* #11366 
* #11367 
* #11381

Spudz76 commented:
No clue why:
`xzcat: /__w/openwrt/openwrt/openwrt/dl/squashfskit-v4.14.tar.xz: No such file or directory`

Tried an alternate source url and still nah.

hauke commented on tools/squashfs4/Makefile:
```diff
@@ -16,26 +16,89 @@ PKG_SOURCE_URL:=https://github.com/plougher/squashfs-tools
+			CONFIG_TOOLS_SQUASHFS_LZO	\
+			CONFIG_TOOLS_SQUASHFS_XATTR	\
+			CONFIG_TOOLS_SQUASHFS_XZ	\
+			CONFIG_TOOLS_SQUASHFS_ZSTD
```
I do not like if this gets rebuild every time we change the default squashfs compression.
If you anyway have all the dependencies to build support for all these compression methods, I would prefer if you always activate them.

Spudz76 replied:
I figured the compile-time or size-minimization crew would complain if every method (and dep thereof) were pulled in for nothing.  Maybe nobody cares in the tools/toolchain.  It does get a whole lot simpler without tracking and just building a does-all version.  Adjustment pending...

Spudz76 replied:
Now it just builds with all features.

dkadioglu commented:
This interesting PR cannot be applied anymore on recent master due to conflicts. I have a Edgerouter X SFP, which suffers from https://github.com/openwrt/openwrt/issues/9085. With changing the compression algorithm, there is a chance, that this might be solved finally, e.g. https://github.com/openwrt/openwrt/issues/9085#issuecomment-1528879095. As far as I understand the conversation here, the PR is fine to be merged, right? It would be great, if @Spudz76 could rebase again and if it could be merged. If further testing is necessary, I am willing to test and give feedback.
One question: When this is applied and I change the compression algo to e.g. ZSTD, can I still do a sysupgrade over the currently installed build with XZ as compression algo? Or do I have to do something special?
Thanks for working on this! The issue of SquashFS corruptions is happening since last year again and again and is absolutely not foreseeable. As I can not provoke it, it is difficult to test, unfortunately. Would be great, if this could be solved, finally.

dkadioglu commented:
Any news regarding this?

Spudz76 commented:
I'll fast forward it shortly.

Spudz76 commented:
I forgot it's way worse than a fast-forward considering some other changes to squashfs4 since, which was why I backburnered it.  Will take me longer, but at least I'm actually digging into it now.

aparcar commented:
@Spudz76 hey could you please give this another go or if you can't, do you mind I look into rebasing (parts) of it? I'm specifically interested in ZSTD for x86 images, so this would come in handy! Thanks for you work!

Spudz76 commented:
@aparcar Finally got around to this.  Still crashes on completely unrelated/untouched things in the tests for whatever mysterious reason.  But still "Works For Me" (tm)

Spudz76 commented:
Still very confused as to why two platforms crash on:
`zstd: can't stat /__w/openwrt/openwrt/openwrt/dl/squashfs4-4.6.1.tar.zst : No such file or directory -- ignored `
...while everything else finds the source just fine.

Or even how to debug that.

Spudz76 commented:
Yeah I get those errors and I thought it was just "normal" / unrelated to this patch.  Admittedly I haven't run a vanilla build in quite some time to see what errors are just regular complaints and which are new.  As you note, everything works fine so I wasn't concerned and decided it was all "informational warnings" or "expected failures" that have workarounds or Plan-B's (or everything works anyway).

Similarly I assumed these are "normal":
```
[   26.748521] mtdblock: MTD device 'devinfo' is NAND, please consider using UBI block devices instead.
[   27.017884] mtdblock: MTD device 'devinfo' is NAND, please consider using UBI block devices instead.
[   27.708573] mtdblock: MTD device 'devinfo' is NAND, please consider using UBI block devices instead.
[   33.898778] do_page_fault(): sending SIGSEGV to uci for invalid write access to 00000000
[   33.907029] epc = 77dac984 in libc.so[77d7c000+a9000]
[   33.912156] ra  = 77dacd94 in libc.so[77d7c000+a9000]
[   34.190490] do_page_fault(): sending SIGSEGV to uci for invalid write access to 00000000
[   34.198624] epc = 77e5e984 in libc.so[77e2e000+a9000]
[   34.203718] ra  = 77e5ed94 in libc.so[77e2e000+a9000]
```

But could be from other patches I have in my "hax" tree that I build from.  I shall try a vanilla build from `main` and see what sort of boot messages I get on this platform (ramips/mt7621).

Spudz76 commented:
@backslashxx Added kernel patches that eliminate the ubifs "bad magic" spew when the magic is squashfs so that it falls through quietly (unless the magic is actually wrong and not squashfs).

Spudz76 commented:
It appears the entire `CONFIG_MTD_ROOTFS_ROOT_DEV` option which triggers trying ubifs before anything else comes from OpenWRT patches anyway, thus is not even part of upstream (yet?).

The offensive code comes from `target/linux/generic/pending-*/492-try-auto-mounting-ubi0-rootfs-in-init-do_mounts.c.patch`

I am not sure why that's in `pending` and not `hack` though, perhaps it has been sent upstream.

IMO there should be a `CONFIG_UBIFS_ROOT` similar to the options for CIFS or NFS root that can control the inclusion of the UBIFS code in the above patch, rather than the fairly more global  `CONFIG_MTD_ROOTFS_ROOT_DEV` which should only do what it says and check for a MTD partition name `rootfs` and set it as the root filesystem, while remaining agnostic about what filesystem, and not do things like assume/try UBIFS before others without explicit indication by another config item, or by setting `rootfstypes=ubifs,squashfs` or such for the current behavior where warranted.  But that might be used for some device where ubifs root is actually being used and the kernel args cannot be changed.

Also considering some investigation toward a similar PR as this one based on UBIFS since it supports most of the same compression methods including zstd, and might be superior to squashfs anyway although that is unclear (when the rootfs is read-only anyway).  I do think squashfs is still better since the block size can be set quite large which generally improves compression ratios, and it would seem at initial glance that ubifs only works in LEB sizes (128K on my devices NAND flash) although that may be different in read-only mode (since eraseblocks are irrelevant when updates are not allowed anyway).  I have been using 1MB with zstd and according to my previous testing that saves 450917 bytes or 3.44 LEBs, that can then be used by the writable overlay ubifs.  There are some patches floating around on squashfs upstream for even larger compression blocksizes that have potential to save even more.

Spudz76 commented:
Tested UBIFS on my platform, the rootfs was enormous compared to squashfs even with zstd enabled, so that's why almost nothing uses it unless it absolutely must (and most seem to only make a ubifs partition containing the kernel to satisfy the minimums of the bootloader, and then pivot to squashfs from there).

Anyway I implemented the `CONFIG_UBIFS_ROOT` option so now it doesn't even include the code that was patched to be quiet by the previous patches.  And furthermore attempted to find all platforms where there may be an actual rootfs in ubifs form and have set those to re-enable the behavior in case they rely on it.  Everything else will just skip that definitely useless code.

Might split this PR into completely separate ones since now some of the commits could probably stand alone, and apparently the squashfs4 tarball situation will never be fixed so this will never pass checks (still sure it's not my fault - maybe doing so many parallel builds hits some connection/spam limit for git clones against github?).

Spudz76 commented:
I have split the UBIFS related stuff into a [different PR here](https://github.com/openwrt/openwrt/pull/15335) since these work separately.  Building from just this PR will reintroduce the boot-spew, apply the other patches on top of (or before) this to achieve the quiet operation again.

tmoore22 reviewed (commented):
good job, patches apply cleanly

Spudz76 commented:
Fixed some failing checks, but still getting apparent lie:
`zstd: can't stat /__w/openwrt/openwrt/openwrt/dl/squashfs4-4.6.1.tar.zst : No such file or directory -- ignored `

There doesn't seem to be any reason that only a couple platforms would choke on that while all the others (and my own build tree) work perfectly.  I do not know how to trace it since I can't reproduce it other than the github checks seem to always fail.

Neustradamus commented:
@Spudz76: Good job since 2 years about it!

gxcreator commented:
Just in case, tests for different algos on MT6722 at Belkin RT3200:
`root@OpenWrt:/tmp# ./lzbench  -ebrotli/zstd/zlib -t8,8 -o1c4 ./dickens
lzbench 2.1 | GCC 14.3.0 | 64-bit Linux |`


| Compressor name         | Compression| Decompress.| Compr. size | Ratio | Filename |
| ---------------         | -----------| -----------| ----------- | ----- | -------- |
| brotli 1.1.0 -0         |  33.5 MB/s |  41.9 MB/s |     4408386 | 43.25 | ./dickens|
| brotli 1.1.0 -1         |  15.5 MB/s |  43.9 MB/s |     4042051 | 39.66 | ./dickens|
| brotli 1.1.0 -2         |  12.0 MB/s |  47.9 MB/s |     3900742 | 38.27 | ./dickens|
| brotli 1.1.0 -3         |  8.97 MB/s |  48.0 MB/s |     3883024 | 38.10 | ./dickens|
| brotli 1.1.0 -4         |  2.96 MB/s |  45.0 MB/s |     3437458 | 33.73 | ./dickens|
| brotli 1.1.0 -5         |  1.71 MB/s |  49.2 MB/s |     3381416 | 33.18 | ./dickens|
| brotli 1.1.0 -6         |  0.97 MB/s |  47.5 MB/s |     3288816 | 32.27 | ./dickens|
| brotli 1.1.0 -7         |  0.55 MB/s |  46.0 MB/s |     3214081 | 31.53 | ./dickens|
| brotli 1.1.0 -8         |  0.33 MB/s |  45.1 MB/s |     3154087 | 30.95 | ./dickens|
| brotli 1.1.0 -9         |  0.21 MB/s |  44.7 MB/s |     3112490 | 30.54 | ./dickens|
| brotli 1.1.0 -10        |  0.20 MB/s |  45.3 MB/s |     2928272 | 28.73 | ./dickens|
| brotli 1.1.0 -11        |  0.09 MB/s |  46.5 MB/s |     2852501 | 27.99 | ./dickens|
| zstd 1.5.7 -1           |  29.2 MB/s |   136 MB/s |     4261734 | 41.81 | ./dickens|
| zstd 1.5.7 -2           |  18.2 MB/s |  94.6 MB/s |     3864999 | 37.92 | ./dickens|
| zstd 1.5.7 -3           |  9.01 MB/s |  80.2 MB/s |     3663504 | 35.94 | ./dickens|
| zstd 1.5.7 -4           |  8.89 MB/s |  71.3 MB/s |     3584538 | 35.17 | ./dickens|
| zstd 1.5.7 -5           |  7.35 MB/s |  82.1 MB/s |     3529639 | 34.63 | ./dickens|
| zstd 1.5.7 -6           |  5.61 MB/s |  87.1 MB/s |     3418515 | 33.54 | ./dickens|
| zstd 1.5.7 -7           |  4.38 MB/s |  81.3 MB/s |     3345600 | 32.82 | ./dickens|
| zstd 1.5.7 -8           |  3.52 MB/s |  83.0 MB/s |     3308676 | 32.46 | ./dickens|
| zstd 1.5.7 -9           |  3.19 MB/s |  80.5 MB/s |     3282524 | 32.21 | ./dickens|
| zstd 1.5.7 -10          |  1.87 MB/s |  74.4 MB/s |     3217624 | 31.57 | ./dickens|
| zstd 1.5.7 -11          |  1.12 MB/s |  71.4 MB/s |     3175497 | 31.16 | ./dickens|
| zstd 1.5.7 -12          |  1.02 MB/s |  70.8 MB/s |     3168791 | 31.09 | ./dickens|
| zstd 1.5.7 -13          |  0.79 MB/s |  71.9 MB/s |     3122251 | 30.63 | ./dickens|
| zstd 1.5.7 -14          |  0.63 MB/s |  69.9 MB/s |     3084587 | 30.26 | ./dickens|
| zstd 1.5.7 -15          |  0.53 MB/s |  68.8 MB/s |     3057963 | 30.00 | ./dickens|
| zstd 1.5.7 -16          |  0.55 MB/s |  72.8 MB/s |     2924826 | 28.70 | ./dickens|
| zstd 1.5.7 -17          |  0.43 MB/s |  72.3 MB/s |     2895048 | 28.40 | ./dickens|
| zstd 1.5.7 -18          |  0.36 MB/s |  71.0 MB/s |     2864357 | 28.10 | ./dickens|
| zstd 1.5.7 -19          |  0.34 MB/s |  70.8 MB/s |     2849941 | 27.96 | ./dickens|
| zstd 1.5.7 -20          |  0.34 MB/s |  70.9 MB/s |     2849093 | 27.95 | ./dickens|
| zstd 1.5.7 -21          |  0.34 MB/s |  70.9 MB/s |     2849141 | 27.95 | ./dickens|
| zstd 1.5.7 -22          |  0.33 MB/s |  70.8 MB/s |     2849064 | 27.95 | ./dickens|
| zlib 1.3.1 -1           |  13.5 MB/s |  64.5 MB/s |     4585618 | 44.99 | ./dickens|
| zlib 1.3.1 -2           |  11.4 MB/s |  67.0 MB/s |     4389480 | 43.07 | ./dickens|
| zlib 1.3.1 -3           |  7.89 MB/s |  69.3 MB/s |     4192085 | 41.13 | ./dickens|
| zlib 1.3.1 -4           |  7.83 MB/s |  70.0 MB/s |     4095027 | 40.18 | ./dickens|
| zlib 1.3.1 -5           |  4.81 MB/s |  71.0 MB/s |     3949175 | 38.75 | ./dickens|
| zlib 1.3.1 -6           |  3.10 MB/s |  71.7 MB/s |     3871634 | 37.99 | ./dickens|
| zlib 1.3.1 -7           |  2.63 MB/s |  72.0 MB/s |     3858882 | 37.86 | ./dickens|
| zlib 1.3.1 -8           |  2.28 MB/s |  72.0 MB/s |     3854735 | 37.82 | ./dickens|
| zlib 1.3.1 -9           |  2.29 MB/s |  71.9 MB/s |     3854735 | 37.82 | ./dickens|
[Params] cIters=1 dIters=1 cTime=8.0 dTime=8.0 chunkSize=0KB cSpeed=0MB
